Add Parent Page setting to Page Settings in editor

thomashorta opened the pull request:
Fixes #11002 

This PR adds the ability to set the Parent Page via the Page Settings screen inside the editor. 

The approach used here was: 
- Add a new section in the  Page Settings screen for "Page Attributes" and add a setting item inside it called "Parent Page" to keep it consistent with the web editor. 
- Upon clicking this new item, open the already existing `PageParentActivity` for result using the `ActivityLauncher`, which only required a minor refactor in its method signature.
- Upon receiving the result with the new `parentId`, set it to the current post/page, like the other settings.
- Update `PostUtils.postHasEdits` to account for the `parentId` as well.

The `PageParentActivity` is not opened from the `Page Settings` when the device is offline, as that could cause inconsistencies, as pointed out by @ovitrif. Same thing for the `Author` option in the same screen, as it is not able to load the Authors while offline. In this scenario (offline) a snackbar with a "no network" message is shown instead.

For getting the parent title, some changes were also required, since that field is not part of the `PostModel` (see "note" below). The changes needed were:
Fetch the parent page for the post in `EditPostRepository` using `PostStore` whenever the Page Settings screen need it and the title was not retrieved before for that same id. We keep a `ParentPage` instance internal to the Repository acting as a "cache".

> Note: `PostModel` has a field called `parentTitle` but that seems to always return the CURRENT page / post title, not its parent. Upon further investigation this comes directly from the backend response, so I couldn't figure out if this is expected and couldn't rely on this field.

To make sure the "Page Parent" selection screen also loads correctly for newly created pages, a change was done in the `PageParentViewModel` to load the pages for `null` `PageModel`s (`pageId = 0`). This scenario happens when you just create a new page, so you are send to the Editor, and from there you try to set a Page Parent using the new setting. At that point, the page is not even saved locally in the `PageStore` so its ID is 0, but that shouldn't be a problem for setting a parent to the in-memory `PageModel` that is being created. Also, after talking to @ovitrif about the analytics event in the save button for this screen we agreed that it makes sense to skip it for new pages in that state (in-memory).

| Android | Web |
| -------  | ----- |
| ![parent_page_top_level_title](https://user-images.githubusercontent.com/5091503/197286411-717d2828-58a8-4f36-97f9-4ef7dc80d99b.png) | ![image](https://user-images.githubusercontent.com/5091503/195739974-7b9125be-10c0-4185-8436-9bc99e493ad0.png) |

Known pre-existing (likely unexpected) behavior:
While testing my changes I realized that there is a behavior that I'm not sure is documented / expected, which is after making any local changes in a Page (changing anything, including parent in the `Editor` screen, then navigating up to the `Pages` screen without hitting `Update`), if you try to `Set Parent` for that specific page from the `Pages` screen, via the overflow menu in the Page item, it will trigger a full update of that content, not just the `Parent Page`.
This was NOT introduced by this PR, but something I found important to add here, since it's related to Parent Pages.

To test:

0. Open the Wordpress app
1. Select "Pages" in the "Menu" tab
2. Select any existing page
3. Tap the overflow meu (`⋮` - three dots)
4. Select "Page Settings"
   - **Verify** "Parent Page" setting in "Parent Attributes" section
5. Tap the "Parent Page" option
   - **Expect** Parent Page selection screen to show a list of possible parents
6. Tap a new parent from the list
7. Tap the Save button at the top right corner
   - **Verify** you are in the "Page Settings" again and the new page is now selected
8. Hit "Update" to save the changes

After doing that you can verify the change was applied in a few different ways:
- In the "Pages" screen the page you updated will be indented under the new parent page selected
- In the "Pages" screen, follow this steps:
   1. Tap the overflow meu (`⋮` - three dots) for the page you just updated
   2. Tap "Set Parent"
   3. **Verify** the current selection is the new parent
- From the "Pages" screen, follow this steps:
   1. Select the page you just updated
   4. Tap the overflow meu (`⋮` - three dots)
   5. Select "Page Settings"
   6. **Verify** "Parent Page" setting in "Parent Attributes" section has the new parent set

## Regression Notes
1. Potential unintended areas of impact
   - Set Parent Page in the Pages Screen (`PagesFragment`)

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

3. What automated tests I added (or what prevented me from doing so)
   - Check the `parentId` field in `EditPostRespository`
   - Check the `parentTitle` field population and update in `EditPostRepository`
   - Unit tests for basic existing functionality and new changes in the `PageParentViewModel`

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -376,16 +385,7 @@ public void onClick(View view) {
-            mFormatContainer.setVisibility(View.GONE);
-            markAsStickyContainer.setVisibility(View.GONE);
-        }
+        if (getEditPostRepository() != null) hideSpecificViews(rootView, getEditPostRepository().isPage());
```
Thank you for considering this improvement here @thomashorta.

I was thinking ...since inside the private `hideSpecificViews` method we're grabbing references to the container views again, wouldn't it then make sense to keep them as fields of the `EditPostSettingsFragment`  class?

thomashorta replied:
Yeah, that was my first thought as well. I was just worried that since this is being called in the `onCreateView` method, having it depend on fields would make it implicitly dependent on **where** that call should happen, specially when called inside `onCreateView`, and changing the call position inside the method could cause a crash.

For instance:
- Calling the method at the end of `onCreateView` would likely always work as those fields are being defined in the body of that method, prior to the call.
- But if for some reason (refactor or something) that call is moved up to the start of the `onCreateView` body, it would mean those variables are still not declared (`null`) which would either cause a crash or cause the function to not work as expected.

ovitrif replied:
I understand your point @thomashorta 👍 

Seems this is a common problem mostly left unsolved Android (views) code.

Thank you for the explanation 🙇

thomashorta replied:
I ended up doing what you suggested and moving the call to `hideSpecificViews` to the bottom of the `onCreateView` method. It makes it look better by just accessing fields, we just have to be careful about call site, but should be fine.

ovitrif commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1663,12 +1663,14 @@
     <string name="post_settings_status" translatable="false">@string/status_and_visibility</string>
     <string name="post_settings_author">Author</string>
-    <string name="post_settings_categories_and_tags">Categories &amp; Tags</string>
+    <string name="post_settings_categories_and_tags">Categories &amp; Tags</string>JetpackPublicDataTest
```
It was probably not the intention to add this here 👍

thomashorta replied:
Definitely not, probably copy-pasted it by mistake when trying to understand why that `JetpackPublicDataTest` was failing for debug builds, lol. Thanks for catching that!

ovitrif commented on WordPress/src/test/java/org/wordpress/android/ui/posts/EditPostRepositoryTest.kt:
```diff
@@ -329,6 +329,19 @@ class EditPostRepositoryTest {
     }
 
+    @Test
+    fun `reads parentId correctly`() {
```
Thank you for adding the new unit test 🙇

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -471,18 +472,13 @@ private void retryFeaturedImageUpload(@NonNull SiteModel site, @NonNull PostImmu
+        hideSpecificViews(getView(), getEditPostRepository().isPage());
         mExcerptTextView.setText(getEditPostRepository().getExcerpt());
+        // TODO need to figure out how to get the parent title to show here
+        mParentTextView.setText(String.format(Locale.getDefault(), "%d", getEditPostRepository().getParentId()));
```
Since we need to display the title of the parent page I suggest continuing to explore the code in order to find a solution that will enable us to do so.

ovitrif reviewed (commented):
Great work @thomashorta 👍

I've tested the new functionality and it works good with what we have at this point, we can continue to find a solution for displaying the parent page title in the page settings, as you already noticed.

The code and the new unit test also looks good 🎉
Thank you for considering to improve the code with refactoring.

I've added a few comments/suggestions, nothing crucial :)

I noticed there is only one commit for all changes, a good improvement suggestion is to split the changes into smaller commits :)

P.S. I understand this is still a draft and my comments are just early suggestions rather than an actual review 👍

thomashorta committed:
Add Parent Page item to Page Settings in Editor

Adding the section for Page Attributes and the Parent Page field as well inside
it. Right now the added changes also include exposing the parentId in the
EditPostRepository so we can show it in the item for now (ideally it will show
the Parent Page Title), but no interaction is programmed yet.

This commit also refactors a small reusable piece of logic used to hide some
UI elements in case of Pages vs Posts, so the logic was extracted to a method.

A unit test was added to EditPostRepository for the new exposed field.

thomashorta committed:
Add ability to set parent from Page Settings

This hooks the Page Parent setting item in Page Settings to actually call
the PageParentActivity for selecting a new parent ID and returning it to the
caller. It then sets the current page ID with that new parent.

The parentId was also added to PostUtils.postHasEdits when checking if posts
are equal or edits have been made.

A minor refactor in the ActivityLauncher.viewPageParentForResult was needed to
be able to call this method from EditPostSettingsFragment.

thomashorta committed:
Add parentTitle fetching to EditPostRepository

This exposes a new plain String field parentTitle which is fetched whenever
a post is loaded in the EditPostRepository as well as a method to update it
manually whenever the original post.parentId might have changed, such as after
a update or updateAsync call setting a new parentId.

Several other classes and unit tests needed to be updated to avoid breaking
the test suite / compilation, since a new parameter was needed in the
loadPostByLocalPostId method.

thomashorta committed:
Display and update the parentTitle in EditPostSettingsFragment

With the new parentTitle field available, it is now shown in the Parent Page
setting field, making sure to also call the updateParentTitle method if the
user changes the current parent.

thomashorta committed:
Add unit tests in EditPostRepository for parentTitle

thomashorta committed:
Fix selecting parent for newly created pages in editor

Newly created pages, which were not even saved locally nor uploaded yet, have
page ID = 0, which results in a null PageModel in PageParentViewModel loading.

A null check was removed to properly list all available parent pages, making
sure to also skip the save analytics event tracking for those scenarios, as all
new pages would have id = 0 and PageModel = null, making the even not relevant.

thomashorta committed:
Add some unit tests to PageParentViewModel

Add enough unit tests to PageParentViewModel class to cover the changes I made
in that class, to avoid future regressions. The tests added here don't cover
everything on the class though, just the newly introduced changes and a few
other small simple scenarios.

thomashorta committed:
Merge branch 'trunk' of github.com:thomashorta/WordPress-Android into issue/11002-page-settings-parent

thomashorta committed:
Update Release Notes for PR 17329

thomashorta commented:
@ovitrif this PR should be ready for review now, based on our conversations and the current state of the code.

thomashorta committed:
Use constant EMPTY_STRING for initializing parentTitle field

thomashorta committed:
Remove PageStore dependency from EditPostRepository

In the end, this was not needed as PAGES are POSTS, so I can just use the
PostStore, which is already present in this class, to fetch the PostModel
that represents the parent, and get the title from it. No need for a PageModel.

Update unit tests accordingly!

thomashorta committed:
Merge branch 'trunk' of github.com:thomashorta/WordPress-Android into issue/11002-page-settings-parent

ovitrif commented on RELEASE-NOTES.txt:
```diff
@@ -2,6 +2,7 @@
 
 21.1
 -----
+* [**] Added parent page to the page settings in the editor and the page parent can be now changed from it. [https://github.com/wordpress-mobile/WordPress-Android/pull/17329]
```
Thank you for thinking of our users and adding a release note for this enhancement.

It might be just me that I find this sentence a bit difficult to read, @thomashorta do you mind checking if it can be improved?

I'm thinking something like "Added functionality to set the parent of a page from page settings in the editor." could work.

P.S. This is going to be reviewed and adjusted by an Editorial team, but I just want to make sure they understand the message clearly 👍

thomashorta replied:
Done!

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostRepository.kt:
```diff
@@ -95,6 +98,10 @@ class EditPostRepository
         get() = post!!.dateLocallyChanged
+    val parentId: Long
+        get() = post!!.parentId
+    var parentTitle: String = EMPTY_STRING
```
I wonder if you've considered the alternative of using a method that gets called from the UI for retrieving the parent page title instead of holding it as a field that gets updated in this repository and what's your thoughts on this.

P.S. This was meant to highlight only the `parentTitle` field but for some reason GitHub didn't apply my late changes to the comment 🤷 .

thomashorta replied:
Yes, it was my first idea! Unfortunately the `parentTitle` field is called inside a `refreshViews` method in `EditPostSettingsFragment`, and that method is called from a LOT of different places (including the `EditPostActivity`) so the reasoning behind creating the field here was to make a kind of memory cache for that value so we can avoid multiple calls to the underlying `PostStore` whenever the view is populated (which seems more costly, being an IO operation).

I spent some time thinking how to make it more transparent to callers (like, not having to manually call `updateParentTitle`) but all solutions I started writing came back to this anyway.

ovitrif replied:
> Yes, it was my first idea!

Appreciated!

>  Unfortunately the parentTitle field is called inside a refreshViews method in EditPostSettingsFragment, and that method is called from a LOT of different places [...]

I understand, and I had the same concerns 👍 

On the other hand it would have removed the need to refactor `loadPostByLocalPostId` with the new `site` parameter, eventually making the impact of the changes smaller in code.

But I do agree the performance cost would be too high, thanks a lot for your feedback on this 🙇

thomashorta replied:
I was able to find a way of refactoring this to make more sense and avoid introducing the `site` parameter`. Check it out and share your thoughts, I think it's looking better indeed, thanks for that comment!

ovitrif replied:
> I was able to find a way of refactoring this to make more sense and avoid introducing the site parameter`. Check it out and share your thoughts, I think it's looking better indeed, thanks for that comment!

You're welcome! I find the new solution to be better, thank you for adding this improvement 🙇 .

While the approach you've taken in the refactoring feels better to me, what I wonder over one detail is if it wouldn't then make sense to use the existing `PageModel` from FluxC instead of defining another page model.

thomashorta replied:
Hmmm, that's a good call, although I will have to use a `PostModel` since I'm not fetching pages anymore, but for this case, since I'm only interested in id and title, it should be good, just a slightly bigger memory footprint compared to the data class I created, but better for consistency! Good call!

thomashorta replied:
Done!

ovitrif replied:
> I will have to use a `PostModel` since I'm not fetching pages anymore, but for this case, since I'm only interested in id and title, it should be good, just a slightly bigger memory footprint compared to the data class I created, but better for consistency! Good call!

This approach makes sense since it aligns with the recent updates in the implementation! I appreciate the decision you've taken and how you adapted the suggestion to the current code.
I too think you've made a good call here 🙇  🎉

P.S. Feel free to mark this as resolved 🙇

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostRepository.kt:
```diff
@@ -210,24 +217,36 @@ class EditPostRepository
 
+    fun updateParentTitle(site: SiteModel) {
+        runBlocking {
+            parentTitle = post?.parentId
```
Nice usage of the Kotlin language features 👏

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -399,6 +410,7 @@ public void onClick(View view) {
             }
         });
 
+        if (getEditPostRepository() != null) hideSpecificViews(getEditPostRepository().isPage());
```
Small nitpick:
Maybe we can align this with the common used pattern in this file and add braces for the if statement.

thomashorta replied:
Sure, will do that change!

I usually try to follow the general Kotlin [style guides for braces](https://developer.android.com/kotlin/style-guide#braces) and end up having those single line ifs some times, lol, but I am happy to change to the pattern used in the file.

ovitrif replied:
Yeah, I would also normally not ask for such changes since I'm also usually in favor of the Kotlin style guides 😅; but for guidelines priority rank higher respecting the patterns specific to a context (in this situation a Java file that already has enough code with its own rules).
Thank you 🙇

thomashorta replied:
Done!

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostRepository.kt:
```diff
@@ -95,6 +98,10 @@ class EditPostRepository
         get() = post!!.tagNameList
     val dateLocallyChanged: String
         get() = post!!.dateLocallyChanged
+    val parentId: Long
```
This seems to be used only in unit tests.
Do we need it here?

thomashorta replied:
You are right, I was using it at some point inside `updateParentTitle`, but moved away from it in favor of simply using `post?.parentId`, I will get rid of it, thanks!

ovitrif commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageParentViewModel.kt:
```diff
@@ -130,21 +132,20 @@ class PageParentViewModel
     fun onSaveButtonTapped() {
-        trackSaveEvent()
-
+        page?.let { trackSaveEvent(it) } // don't track event for null pages (just created but not yet saved / uploaded)
```
Let's see what feedback we get on this 🤞

ovitrif replied:
Received feedback confirms it's a valid solution. Closing this 👍

ovitrif commented on WordPress/src/test/java/org/wordpress/android/ui/posts/EditPostRepositoryTest.kt:
```diff
@@ -499,6 +553,40 @@ class EditPostRepositoryTest {
+    }
+
+    @Test
+    fun `loads parentTitle from store for post loaded by remote id is 0`() = test {
```
The name of this test can be improved imho, I find the title of [this test similar to it](https://github.com/wordpress-mobile/WordPress-Android/pull/17329/files#diff-edb47136a3f8bef313720de26c2e70f0a4ba92239884f7dc0545a4b9a4f08df8R530) a bit cleaner.

thomashorta replied:
Done!

ovitrif commented on WordPress/src/test/java/org/wordpress/android/ui/posts/EditPostRepositoryTest.kt:
```diff
@@ -522,4 +610,33 @@ class EditPostRepositoryTest {
+
+        assertThat(editPostRepository.parentTitle).isEqualTo("")
+    }
 }
```
Thank you for adding these unit tests @thomashorta 🙇

ovitrif commented on WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageParentViewModelTest.kt:
```diff
@@ -0,0 +1,186 @@
+
+@InternalCoroutinesApi
+@RunWith(MockitoJUnitRunner::class)
+class PageParentViewModelTest {
```
Many thanks for introducing unit tests for the `PageParentViewModel` 🙇

ovitrif reviewed (commented):
Thank you for your progress on this enhancement @thomashorta 🙇 

I've reviewed the changes and tested, overall things are looking good 🎉 

I added a few comments for consideration though nothing big.

I also noticed an issue with setting the parent page, for local-only pages when internet connectivity is not available. It can be reproduced with the following steps:

1. Open app & go to Pages List
2. Disable internet connectivity on the device
3. Create a new blank page (since it's the only option)
4. Publish the page
5. Create another new page
6. Go to Page settings > Parent Page
7. Select the page created at step 3 (local page) to be parent
8. Tap Save
9. **Notice** under Page parent the parent title displays as "Not Set".
10. Going back to "Set Parent" the previous selection is lost.

<details>
<summary>Recording of the issue</summary>

https://user-images.githubusercontent.com/4588074/196797823-675357f9-3d5b-4235-93bb-332469998334.mp4
</details>

It's not possible to run into this issue when trying to set parent from the pages list because the screen is not opening without internet connection.
This could be a solution also for the set parent flow via the page settings 👍 

And last but not least I would suggest to ask for design feedback on this enhancement proposal since we're close to getting ready with the implementation 🎉

thomashorta committed:
Update RELEASE-NOTES for better readability (PR comment)

thomashorta committed:
Align if clause with file pattern (PR comment)

thomashorta committed:
Remove parentId field from EditPostRepository (PR comment)

thomashorta committed:
Improve test name for readability (PR comment)

Change the name of one of the unit tests to make it consistent
with a similar test and the rest of the test names.

thomashorta committed:
Get parentTitle from store only when needed (PR comment)

Improved the code to only fetch the parent title for the current parentId
when a caller asks for getParentTitle AND we know that the parentID changed,
storing the current <parentId, parentTitle> pair as a ParentPage internal
object in EditPostRepository.

This avoids introducing the siteModel in the loadPostByLocalPostId parameters
and makes more sense from a caller perspective (no need to know when to update
the title).

ravishanker commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java:
```diff
@@ -99,12 +101,15 @@ public class EditPostSettingsFragment extends Fragment {
     private SiteSettingsInterface mSiteSettings;
 
-    private LinearLayout mCategoriesContainer;
+    private LinearLayout mCategoriesTagsContainer;
```
If there's time.
- Would be good to refactor this to Kotlin and use Hilt to reduce the number of lines drastically (there are hundreds of lines that are layout related!)
- Since there's already a ViewModel, All the repository related stuff could go there

wdyt?

thomashorta replied:
I think this refactor might be a risky move because the `Repository` exposes `PostModel` fields directly and that underlying `post` is being updated from more than one place (i.e.: `EditPostActivity`, `EditPostSettingsFragment`, `PublishSettingsViewModel`, and more) so ensuring proper data propagation might be complex and very likely to introduce bugs with the time I have to work on this. :sweat_smile:

Also, the `EditPostPublishSettingsViewModel` is used in a few other Fragments as well and it looks like it is specific to Publish Settings at the moment, so it would get a hit on single responsibility if I were to reuse that.

Unfortunately, I think it's safer to not do the refactor right now, although it would be a nice improvement in the future. Does that sound good?

ovitrif replied:
I really like the suggestion by @ravishanker but I tend to agree with your remarks Thomas and I'm a little concerned we won't have enough time to progress significantly on a refactoring.

thomashorta committed:
Change empty page parent label to say Top level

As suggested by Thomas Bishop, showing "Not set" for the Parent Page setting
might be confusing to our users, so setting it to "Top level" makes more sense.

thomashorta committed:
Use PostModel for parent page in EditPostRepository

Instead of using another page model class (ParentPage), we should leverage the
existing PostModel object, simplifying the logic and keeping things consistent.

thomashorta committed:
Show "no network" snackbar when tapping some settings

A couple of settings don't work properly when network is not available so this
commit adds a "guard", preventing those actions to continue and showing a
snackbar with a "no network" message, similar to what the PagesFragment has.

The settings affected are:
- Parent Page
  To avoid the user selecting an invalid page as parent (that is not yet
  published remotely), it's better to disable the access to this option while
  offline to avoid issues.
- Author
  The author list doesn't load properly while offline, showing a Dialog with
  a "Loading..." message.

thomashorta committed:
Don't show local-only pages in Page Parent selection list

Even though the PageParent screen is not called when the device is offline,
it's better to put an additional safe-guard in case this page is opened when
there are pages that are local-only (created while device was offline and not
published remotely yet), as they cannot be used as parent to other pages. So
the changes here filter out local-only pages (the ones that have negative
remote IDs).

thomashorta commented:
@ovitrif:
> I also noticed an issue with setting the parent page, for local-only pages when internet connectivity is not available.

As we talked earlier and you suggested, I added code to prevent the `ParentPageActivity` from being opened from the `EditPostSettingsFragment` while the device is offline. I noticed the `Author` dialog also doesn't work properly while offline, so I added that improvement to that setting as well. It now doesn't let the action go through and shows a snackbar explaining there's no network (like what happens in `PagesFragment`).

I also implemented a second safe-guard in the `ParentPageViewModel` so even if that screen is reached, pages that are local-only (never had been published remotely yet) are NOT shown in the parent list, as they are NOT valid parents in that state.

ovitrif commented:
> @ovitrif:
> As we talked earlier and you suggested, I added code to prevent the `ParentPageActivity` from being opened from the `EditPostSettingsFragment` while the device is offline. I noticed the `Author` dialog also doesn't work properly while offline, so I added that improvement to that setting as well. It now doesn't let the action go through and shows a snackbar explaining there's no network (like what happens in `PagesFragment`).

Thank you for tackling this issue @thomashorta. The code changes you've applied to handle this case look good to me, it also works as expected in tests 🎉 

> I also implemented a second safe-guard in the `ParentPageViewModel` so even if that screen is reached, pages that are local-only (never had been published remotely yet) are NOT shown in the parent list, as they are NOT valid parents in that state.

Great thinking here! Actually this fix was needed, otherwise considering this scenario we'd still see local pages to set as parent, and possibly run into issues:
If there are many local pages with media in the upload queue (since they were created while offline), if we enable internet connection when looking at page settings for one of the new pages, then we tap `Parent Page`, the list would have shown pages that still local-only and queued for upload.

This is best tested with a slow internet connection, when not all pages are yet uploaded and we land on the `Set Parent` screen.

Thank you for proactively adding this fix 🙇

ovitrif commented:
> Same thing for the Author option in the same screen, as it is not able to load the Authors while offline. In this scenario (offline) a snackbar with a "no network" message is shown instead.

Many thanks for handling this case too @thomashorta 🙇 .

ovitrif reviewed (approved):
Works as expected with the latest change 🎉

thehenrybyrd commented:
@thomashorta thanks for including parent page in the editor page settings!

In your PR description, you mention this issue you ran into:

> While testing my changes I realized that there is a behavior that I'm not sure is documented / expected, which is after making any local changes in a Page (changing anything, including parent in the Editor screen, then ....
This was NOT introduced by this PR, but something I found important to add here, since it's related to Parent Pages.

Nice find, and thanks for mentioning it here - definitely important to document! Can you create a new issue for that so we don't loose track?

thomashorta commented:
>Nice find, and thanks for mentioning it here - definitely important to document! Can you create a new issue for that so we don't loose track?

Hi @thehenrybyrd I just created the ticket with that issue / improvement opportunity: #17417
